integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/management/CarbonInsertFromStageCommand.scala

```diff
@@ -269,11 +281,90 @@ case class CarbonInsertFromStageCommand(
       SparkSQLUtil.sessionState(spark).newHadoopConf()
     ).map { row =>
         (row._1, FailureCauses.NONE == row._2._2.failureCauses)
-      }
+    }
+
+    LOGGER.info(s"finish data loading, time taken ${System.currentTimeMillis() - start}ms")
+  }
+
+  /**
+   * Start global sort loading of partition table
+   */
+  private def startLoadingWithPartition(
+      spark: SparkSession,
+      table: CarbonTable,
+      loadModel: CarbonLoadModel,
+      stageInput: Seq[StageInput]
+    ): Unit = {
+    val partitionDataList = listPartitionFiles(stageInput)
+    val start = System.currentTimeMillis()
+    var index = 0
+    partitionDataList.map(
+      partitionData => {
+        index = index + 1
+        val partition = partitionData._1
+        val splits = partitionData._2
+        LOGGER.info(s"start to load ${splits.size} files into " +
+          s"${table.getDatabaseName}.${table.getTableName}")
+        val dataFrame = createInputDataFrameOfInternalRow(spark, table, splits)
+        val columns = dataFrame.columns
+        val header = columns.mkString(",")
+        val selectColumns = columns.filter(!partition.contains(_))
+        val selectedDataFrame = dataFrame.select(selectColumns.head, selectColumns.tail : _*)
 
+        val loadCommand = CarbonLoadDataCommand(
+          databaseNameOp = Option(table.getDatabaseName),
+          tableName = table.getTableName,
+          factPathFromUser = null,
+          dimFilesPath = Seq(),
+          options = scala.collection.immutable.Map("fileheader" -> header),
+          isOverwriteTable = false,
+          inputSqlString = null,
+          dataFrame = Some(selectedDataFrame),
+          updateModel = None,
+          tableInfoOp = None,
+          internalOptions = Map.empty,
+          partition = partition
+        )
+        loadCommand.run(spark)
+      }
+    )
     LOGGER.info(s"finish data loading, time taken ${System.currentTimeMillis() - start}ms")
   }
 
+  private def listPartitionFiles(stageInputs : Seq[StageInput]):
+  Seq[(Map[String, Option[String]], Seq[InputSplit])] = {
+    val partitionMap = new util.HashMap[Map[String, Option[String]], util.List[InputSplit]]()
+    stageInputs.foreach(
+      stageInput => {
+        val locations = stageInput.getLocations.asScala
+        locations.foreach(
+          location => {
+            val partition = location.getPartitions.asScala.map(t => (t._1, Option(t._2))).toMap
```

Comment:
Why Option is required? Can't we remove it?

Reply:
The CarbonLoadDataCommand required Option, so we cann't remove it here.